Xamarin.Forms.Build.Tasks/ModuleDefinitionExtensions.cs

```diff
@@ -1,32 +1,182 @@
+		{
+			var method = module
+				.ImportReference(type)
+				.ResolveCached()
+				.Methods
+				.FirstOrDefault(md =>
+								   !md.IsConstructor
+								&& md.Name == methodName
+								&& md.Parameters.Count == paramCount
+								&& (predicate?.Invoke(md) ?? true));
+			if (method is null)
+				return null;
+			var methodRef = module.ImportReference(method);
+			if (classArguments == null)
+				return methodRef;
+			return module.ImportReference(methodRef.ResolveGenericParameters(type.MakeGenericInstanceType(classArguments), module));
+		}
 
-		public static MethodReference ImportReferenceCached(this ModuleDefinition module, System.Reflection.MethodBase method)
+		public static MethodReference ImportMethodReference(this ModuleDefinition module,
+															(string assemblyName, string clrNamespace, string typeName) type,
+															string methodName, int paramCount, Func<MethodDefinition, bool> predicate = null,
+															(string assemblyName, string clrNamespace, string typeName)[] classArguments = null)
 		{
-			var key = new Tuple<ModuleDefinition, System.Reflection.MethodBase>(module, method);
-			if (MbMr.TryGetValue(key, out var result))
-				return result;
-			return MbMr[key] = module.ImportReference(method);
+			return module.ImportMethodReference(module.GetTypeDefinition(type), methodName, paramCount, predicate,
+												classArguments?.Select(gp => module.GetTypeDefinition((gp.assemblyName, gp.clrNamespace, gp.typeName))).ToArray());
+		}
+
+		static Dictionary<(ModuleDefinition module, (string assemblyName, string clrNamespace, string typeName)), TypeDefinition> typeDefCache
+			= new Dictionary<(ModuleDefinition module, (string assemblyName, string clrNamespace, string typeName)), TypeDefinition>();
+
+		public static TypeDefinition GetTypeDefinition(this ModuleDefinition module, (string assemblyName, string clrNamespace, string typeName) type)
+		{
+			if (typeDefCache.TryGetValue((module, type), out TypeDefinition cachedTypeDefinition))
+				return cachedTypeDefinition;
+
+			var asm = module.Assembly.Name.Name == type.assemblyName
+							? module.Assembly
+							: module.AssemblyResolver.Resolve(AssemblyNameReference.Parse(type.assemblyName));
+			var typeDef = asm.MainModule.GetType($"{type.clrNamespace}.{type.typeName}");
+			if (typeDef != null) {
+				typeDefCache.Add((module, type), typeDef);
+				return typeDef;
+			}
+			var exportedType = asm.MainModule.ExportedTypes.FirstOrDefault(
+				arg => arg.IsForwarder && arg.Namespace == type.clrNamespace && arg.Name == type.typeName);
+			if (exportedType != null) {
+				typeDef = exportedType.Resolve();
+				typeDefCache.Add((module, type), typeDef);
+				return typeDef;
+			}
+
+			//I hate you, netstandard
+			if (type.assemblyName == "mscorlib" && type.clrNamespace == "System.Reflection")
+				return module.GetTypeDefinition(("System.Reflection", type.clrNamespace, type.typeName));
+				throw new Exception($"Failed to get typedef for {type}");
```

Comment:
👮 This LOOKS like this exception happens as part of this `if` clause because of the indentation.